ccan/ccan/graphql/graphql.c

```diff
@@ -0,0 +1,1031 @@
+	INIT(directive);
+	parse_punct(ARGS, '@'); REQ
+	obj->name = parse_name(ARGS); REQ
+	obj->args = parse_arguments(ARGS); OPT
+	EXIT;
+}
+
+
+/* The following functions construct the "leaves" of the abstract syntax tree. */
+
+RET parse_keyword(PARAMS, const char *keyword, const char *errmsg) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'a') {
+		*err = errmsg; return NULL;
+	}
+	if (!streq(tok->token_string, keyword)) {
+		*err = errmsg; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+// Note: a static buffer is used here.
+RET parse_punct(PARAMS, int punct) {
+	static char punctbuf[16];
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != punct) {
+		if (punct == 0x2026)
+			sprintf(punctbuf, "expected: '...'");
+		else
+			sprintf(punctbuf, "expected: '%c'", punct);
+		*err = punctbuf; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_name(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'a') {
+		*err = "name expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_int(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'i') {
+		*err = "integer expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_float(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 'f') {
+		*err = "float expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+RET parse_string(PARAMS) {
+	struct graphql_token *tok = list_top(tokens, struct graphql_token, list);
+	if (!tok || tok->token_type != 's') {
+		*err = "string expected"; return NULL;
+	}
+	CONSUME_ONE;
+	return tok;
+}
+
+
+// Convert input string into tokens.
+const char *graphql_lex(const char *input, const tal_t *ctx, struct list_head **tokens) {
+
+	unsigned int c;
+	const char *p, *line_beginning;
+	unsigned int line_num = 1;
+	struct list_head *tok_list;
+	struct graphql_token *tok;
+
+	// Initialize token output list.
+	tok_list = tal(ctx, struct list_head);
+	if (tokens)
+		*tokens = tok_list;
+	list_head_init(tok_list);
+
+	// Note: label and goto are used here like a continue statement except that
+        // it skips iteration, for when characters are fetched in the loop body.
+	p = input;
+	line_beginning = p;
+	do {
+		c = *p++;
+newchar:
+		// Consume line terminators and increment line counter.
+		if (LINE_TERMINATOR(c)) {
+			unsigned int c0 = c;
+			c = *p++;
+			if (c0 == 10 || c0 == 13)
+				line_num++;
+			if (c0 == 13 && c == 10)
+				c = *p++;
+			line_beginning = p - 1;
+			goto newchar;
+		}
+
+		// Consume other ignored tokens.
+		if (COMMA(c) || WHITE_SPACE(c)) {
+			c = *p++;
+			goto newchar;
+		}
+		if (COMMENT(c)) {
+			while ((c = *p++) != EOF && !EOF_CHAR(c) && COMMENT_CHAR(c))
```

Comment:
EOF check is wrong here, it's a nul-terminated string right?

Reply:
Yes, good catch. (In an early incarnation, this function was using fgetc which of course can return EOF.)